notebook-api/notebook/notebook.py

```diff
@@ -0,0 +1,434 @@
+                  (event['type'], event['object'].kind, event['object'].metadata.name))
+
+            if event['type'] == 'MODIFIED' or event['type'] == 'ADDED':
+                ws.send(
+                    f'{{"event": "{event["type"]}", "resource": {marshall_json([obj], pod_paths, True)}}}')
+
+            if event['type'] == 'DELETED':
+
+                ws.send(
+                    f'{{"event": "DELETED", "resource": {marshall_json([obj], pod_paths, True)}}}')
+        except Exception as e:
+            log.error(e)
+            return
+
+    while not ws.closed:
+        gevent.sleep(10)
+
+
+@app.route('/api/verify/<svc_name>/', methods=['GET'])
+def verify(svc_name):
+    access_token = request.args.get('authorization')
+    token = request.args.get('token')
+
+    if not access_token:
+        return forbidden()
+
+    resp = requests.get(f'{AUTH_GATEWAY}/verify',
+                        headers={'Authorization': f'Bearer {access_token}'})
+
+    if resp.status_code != 200:
+        return forbidden()
+
+    user_id = resp.headers.get('User')
+
+    if not user_id:
+        return forbidden()
+
+    res = k8s.read_namespaced_service(svc_name, 'default')
+
+    l = res.metadata.labels
+    if l['jupyter_token'] != token or l['user_id'] != UNSAFE_user_id_transform(user_id):
+        return forbidden()
+
+    return '', 200
+
+
+# read_svc_status checks that the pod's backing service exists
+# It returns a string status, which must match the corresponding Pod status
+# Ex: If Kubernetes names running pod status (phase) "Running", and svc is running
+# we should return "Running" here
+# TODO: Inspect exceptions to distinguish between 404 and others
+
+def read_svc_status(svc_name):
+    try:
+        svc = k8s.read_namespaced_service(svc_name, 'default')
+        return 'Running'
+    except:
+        return 'Deleted'
+
+
+# TODO: support lambda expressions
+# TODO: had default key as path name, but this adds branches/if statements
+# in tight loop, dont think the space savings worth it
+
+pod_paths = [
+    ['name', 'metadata.labels.name'],
+    ['pod_name', 'metadata.name'],
+    ['svc_name', 'metadata.labels.svc_name'],
+    ['pod_status', 'status.phase'],
+    ['svc_status', 'metadata.labels.svc_name', lambda x: read_svc_status(x)],
+    ['creation_date', 'metadata.creation_timestamp',
+        lambda x: x.strftime('%D')],
+    ['token', 'metadata.labels.jupyter_token']
+]
+
+
+@app.route('/api', methods=['GET'])
+def get_notebooks():
+    user_id = request.headers.get('User')
+
+    if not user_id:
+        return forbidden()
+
+    # svcs = k8s.list_namespaced_service(
+    #     namespace='default',
+    #     watch=False,
+    #     label_selector='user_id=' + UNSAFE_user_id_transform(user_id),
+    #     _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS).items
+    # k8s.read_namespaced_service(svc_name, 'default')
+    pods = k8s.list_namespaced_pod(
+        namespace='default', watch=False,
```

Comment:
The default is to not watch so let's not set it here. We should also include a `_request_timeout`. That's an undocumented kwarg that I dug up a while back to set a request timeout. Unfortunately, if you don't set it, swagger never times out.

Reply:
Got it, will add it. Thanks for note on watch.

Reply:
Btw, I just saw "timeout_seconds" for list_namespaced_pod. It seems they've finally exposed this officially?

https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/CoreV1Api.md#list_namespaced_pod

Maybe in the transition out of beta2